### DISCUSSION TITLE
Configuration questions

### DISCUSSION THREAD

**Discussion by Clive321A** [FortuneN/FineCodeCoverage#366] [Q&A] [answered]
Hi, struggling with configuration, any help appreciated

1) Id like to exclude certain csproj files.
2) Use MS Code coverage if available in VS2022 Professional (I thought that was only enterprise)
3) Stop other NuGet packages displaying in output, ie MOQ,  Microsoft.TestPlatform are displaying with 0 covered, and errors in logs saying it cant find the source for these. (ie  File '/_/src/Moq/SetupWithOutParameterSupport.cs' does not exist (any more).)

To exclude I have tried FCCExcludeFromCodeCoverage/ in csproj,  
Tried putting a finecodecoverage-settings.xml in the root of the project with FALSE for enabled.
Tried, setting Exclude options in .runsettings (Which I got from the temporary one the FCC creates)
Tried setting the Fine Code Coverage options in Tools/Options in VS, and trying to set the "ExcludeByFile" and/or "Exclude" with various wildcards.
Tried putting a PropertyGroup Label="FineCodeCoverage" into the csproj with Enabled false

I have read over the setup guide lots of times, in case im missing something.

**Comment by tonyhallett** [comment]
> To exclude I have tried FCCExcludeFromCodeCoverage/ in csproj,

This needs to be placed in a csproj that is directly referenced by your test project - see https://github.com/FortuneN/FineCodeCoverage/issues/363

> Tried putting a finecodecoverage-settings.xml in the root of the project with FALSE for enabled.
> Tried putting a PropertyGroup Label="FineCodeCoverage" into the csproj with Enabled false

This will works for non ms code coverage and conditionally for ms code coverage.
If you are supplying your own runsettings file we will intercept it and set the DataCollector enabled attribute.  If you are providing a single runsettings for all projects in a solution then the best that we can do is disable if all projects are disabled. 

Also, if you are using property group FineCodeCoverage it needs to be in the test project.
 

> Tried, setting Exclude options in .runsettings (Which I got from the temporary one the FCC creates)

Did you tell visual studio about your .runsettings https://learn.microsoft.com/en-us/visualstudio/test/configure-unit-tests-by-using-a-dot-runsettings-file?view=vs-2022#specify-a-run-settings-file-in-the-ide

I would use FCC auto generated runsettings unless you have a specific need otherwise you will have to specify replacements in your runsettings. For example
```
 <ModulePaths>
	<Exclude>
		%fcc_modulepaths_exclude%
	</Exclude>
</ModulePaths>
```

> Tried setting the Fine Code Coverage options in Tools/Options in VS, and trying to set the "ExcludeByFile" and/or "Exclude" with various wildcards.
ExcludeByFile and Exclude are not ms code coverage.

https://learn.microsoft.com/en-us/visualstudio/test/customizing-code-coverage-analysis?view=vs-2022#include-or-exclude-assemblies-and-members 
https://learn.microsoft.com/en-us/visualstudio/test/customizing-code-coverage-analysis?view=vs-2022#regular-expressions

Also available to you is the [ExcludeFromCodeCoverageAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.excludefromcodecoverageattribute?view=net-8.0) which can be applied to a whole assembly - but not the case with [Net Framework](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.excludefromcodecoverageattribute?view=netframework-4.8.1)

**Comment by Clive321A** [comment]
Great thanks Tony, was that I didnt have a reference to the project in my test project, only a indirect reference. They are being ignored now by using the FCCExcludeFromCodeCoverage in csproj

The only outstanding thing is why other nuget packages are being included, like moq for example, I dont understand why just these nuget packages are, and not all the other ones.

![image](https://github.com/FortuneN/FineCodeCoverage/assets/29695733/72fe7775-3872-4128-8a97-2c07c1bc9176)

**Reply by tonyhallett** [reply, marked as answer]
If they provide pdbs they will be included. Just use the ModulePath Exclude regular expression to exclude them.

**Reply by Clive321A** [reply]
Excellent, thanks have added reference to dll into module exclude for each and they are gone..
